### ISSUE TITLE
What's up with designspace files that point to UFOs whose feature files dn't match?

### ISSUE THREAD

**Issue by rsheeter** [googlefonts/fontmake-rs#72]
According to https://github.com/rsheeter/google_fonts_sources/blob/main/ufo_fea_consistency.py sometimes a designspace references many feature files that match ... and sometimes not. I'm not quite sure what to make of the case where they don't match. Maybe I just oversimplified and failed to skip some I should have skipped?

https://gist.github.com/rsheeter/31f5233580a4328c8c3ac46d27646b2b shows results as of 1/6/2023

**Comment by rsheeter** [comment]
glyphsLib has code that complains when this occurs; https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/Lib/glyphsLib/builder/features.py#L327

**Comment by rsheeter** [comment]
Expert feedback (@simoncozens, @behdad) suggests GPOS rules can and should vary between masters, GSUB differing we don't support, either taking the main master GSUB only or requiring them to be identical (TODO: check which)

**Comment by simoncozens** [comment]
So is this:

```
regular.ufo/features.fea: feature kern { pos A B 50; pos A C 100; } kern;
bold.ufo/features.fea: feature kern { pos A B 75; } kern; # pos A C 0 is implied, and filled in by the merger
```

**Comment by anthrotype** [comment]
> taking the main master GSUB only or requiring them to be identical (TODO: check which).

I'd say the latter. In the current fontmake pipeline, we build each UFO master's features.fea one by one (they may or not have GSUB features defined), then fontTools.varLib runs the VariationMerger.mergeTables on all GDEF, GPOS and GSUB tables, but currently this only handles variations in GDEF and GPOS, not also inside GSUB; as a result, if masters' GSUB turn out to be different from one another, an error would occur at merging time. So effectively GSUBs they have to be identical at the binary level (it doesn't matter if a lookup is called FooA or FooB at that stage as long as they do the same thing).

**Comment by anthrotype** [comment]
> GPOS rules can and should vary between masters

indeed, but I'd say the majority of those differences are auto-generated kern/mark/mkmk (GPOS) features or (GDEF) ligature carets, as generated by ufo2ft feature writers. However these won't be visible by simply looking at the source files, as they are generated at build time (from kerning.plist or glyph anchors definitions) before the features get compiled with feaLib.
